exploration/formatted-parts.md

```diff
@@ -0,0 +1,212 @@
+Recent examples of web browsers needing to account for such user behaviour are available from
+[June 2022](https://github.com/WebKit/WebKit/commit/1dc01f753d89a85ee19df8e8bd75f4aece80c594) and
+[November 2022](https://bugs.chromium.org/p/v8/issues/detail?id=13494).
+
+## Use-Cases
+
+- Markup elements
+- Non-string values
+- Message post-processors
+- Decoration of placeholder interior parts.
+  For example, identifying the separate fields in these two currency values
+  (notice that the symbol, number, and fraction fields
+  are not in the same order and that the separator has been omitted):
+  ![image](https://github.com/unicode-org/message-format-wg/assets/69082/cb68c87f-9c0c-4bc6-b9a0-b1f97b2b789a)
+  ![image](https://github.com/unicode-org/message-format-wg/assets/69082/aedd4e66-7d47-4026-8b93-4ba061bb4d84)
+- Supplying bidirectional isolation of placeholders,
+  such as by using HTML's `span` element with a `dir` attribute
+  based on the direction of the placeholder.
+
+## Requirements
+
+- Define an iterable sequence of formatted part objects.
+- Include metadata for each part, such as type, source, direction, and locale.
+- Allow the representation of non-string values.
+- Allow the representation of values that consist of an iterable sequence of formatted parts.
+- Be able to represent each resolved value of a pattern with any number of formatted parts, including none.
+- Define the formatted parts in a manner that allows synonymous but appropriate implementations in different programming languages.
+
+## Constraints
+
+- The JS Intl formatters already include formatted-parts representations for each supported data type.
+  The JS implementation of the MF2 formatted-parts representation should be able to match their structure,
+  at least as far as that's possible and appropriate.
+
+## Proposed Design
+
+The formatted-parts API is included in the spec as an optional but recommended formatting target.
+
+The shape of the formatted-parts output is defined in a manner similar to the data model,
+which includes TypeScript, JSON Schema, and XML DTD definitions of the same data structure.
+
+At the top level, the formatted-parts result is an iterable sequence of parts.
+Parts corresponding to each _text_ can be simpler than those of _expressions_,
+as they do not have a `source` other than their `value`,
+or set any of the other possible metadata fields.
+
+```ts
+type MessageParts = Iterable<
+  MessageTextPart | MessageExpressionPart | MessageBiDiIsolationPart
+>;
+
+interface MessageTextPart {
+  type: "text";
+  value: string;
+}
+```
+
+For MessageExpressionPart, the `source` corresponds to the expression's fallback value.
+The `dir` and `locale` attributes of a part may be inherited from the message
+or from the operand (if present),
+or overridden by an expression attribute or formatting function,
+or otherwise set by the implementation.
+Each part should have at most one of `value` or `parts` defined;
+some may have none.
```

Comment:
While this may be how this gets implemented in certain implementations, I think our definition should be stricter. Something like:

```ts
type MessageExpressionPart = MessageSingleValuePart | MessageMultiValuePart;

interface MessageSingleValuePart {
  type: string;
  source: string;
  value?: unknown;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}

interface MessageMultiValuePart {
  type: string;
  source: string;
  parts?: Iterable<{ type: string; value: unknown; source?: string }>;
  dir?: "ltr" | "rtl" | "auto";
  locale?: string;
}
```

Reply:
This sounds like a sensible way to specify this, should this design doc be accepted. Thought with a required `parts` to explicitly differentiate the two.

Reply:
Added the interfaces.